src/objective/rank_obj.cu

```diff
@@ -0,0 +1,637 @@
+    int num_elems = groups.back();
+
+    dsorted_labels_.resize(num_elems);
+
+    doriginal_pos_.resize(num_elems);
+    thrust::sequence(doriginal_pos_.begin(), doriginal_pos_.end());
+
+    group_segments_.resize(num_elems);
+
+    dgroups_ = groups;
+
+    // Launch a kernel that populates the segment information for the different groups
+    uint32_t *gsegs = group_segments_.data().get();
+    const unsigned *dgroups = dgroups_.data().get();
+    int ngroups = dgroups_.size();
+    dh::LaunchN(device_id_, num_elems, nullptr, [=] __device__(unsigned idx){
+      // Find the group first
+      int group_idx = dh::UpperBound(dgroups, ngroups, idx);
+      gsegs[idx] = group_idx - 1;
+    });
+  }
+
+  // Sort the groups by labels. We would like to avoid using predicates to perform the sort,
+  // as thrust resorts to using a merge sort as opposed to a much much faster radix sort
+  // when comparators are used. Hence, the following algorithm is used. This is done so that
+  // we can grab the appropriate prediction values from the original list later, after the
+  // labels are sorted.
+  //
+  // Here is the internal representation:
+  // dgroups_:          [ 0, 3, 5, 8, 10 ]
+  // group_segments_:   0 0 0 | 1 1 | 2 2 2 | 3 3
+  // doriginal_pos_:    0 1 2 | 3 4 | 5 6 7 | 8 9
+  // dsorted_labels_:   1 0 1 | 2 1 | 1 3 3 | 4 4 (from original labels)
+  //
+  // Sort the labels first and make a note of the original positions in doriginal_pos_
+  // based on the sort
+  // dsorted_labels_:   4 4 3 3 2 1 1 1 1 0
+  // doriginal_pos_:    8 9 6 7 3 0 2 4 5 1
+  // NOTE: This consumes space, but is much faster than some of the other approaches - sorting
+  //       in kernel, sorting using predicates etc.
+  void Sort(const HostDeviceVector<bst_float> &dlabels) {
+    dsorted_labels_.assign(dh::tcbegin(dlabels), dh::tcend(dlabels));
+    thrust::stable_sort_by_key(dsorted_labels_.begin(), dsorted_labels_.end(),
+                               doriginal_pos_.begin(), thrust::greater<float>());
+
+    // Next, gather the segments based on the doriginal_pos_. This is to reflect the
+    // holisitic label sort order on the segments
+    // group_segments_c_:   3 3 2 2 1 0 0 1 2 0
+    // doriginal_pos_:      8 9 6 7 3 0 2 4 5 1 (stays the same)
+    thrust::device_vector<uint32_t> group_segments_c(group_segments_);
+    thrust::gather(doriginal_pos_.begin(), doriginal_pos_.end(),
+                   group_segments_.begin(), group_segments_c.begin());
+
+    // Now, sort the group segments so that you may bring the labels within the group together,
+    // in the process also noting the relative changes to the doriginal_pos_ while that happens
+    // group_segments_c_:   0 0 0 1 1 2 2 2 3 3
+    // doriginal_pos_:      0 2 1 3 4 6 7 5 8 9
+    thrust::stable_sort_by_key(group_segments_c.begin(), group_segments_c.end(),
+                               doriginal_pos_.begin(), thrust::less<int>());
```

Comment:
If you provide a comparison operator doesn't this force thrust to use merge sort instead of radix sort?

Also, use a caching allocator as per the above comment.

Reply:
thrust sorts are optimized for standard binary predicates - greater and lesser; for anything that is not either of those, it switches back to a slower version. it chooses this at compile time.

https://github.com/thrust/thrust/blob/master/thrust/system/cuda/detail/sort.h#L1512
https://github.com/thrust/thrust/blob/master/thrust/system/cuda/detail/sort.h#L1559

i tried to fake it, by reopening the thrust::detail namespace to define a type that equates my own comparator to one of thrust greater/lesser (such that it evaluates to a true_type) earlier to fool's errand, as they are much smarter than that :-)

the 2nd part of your comment has been addressed.